TensorIteratorReduce: Avoid tensor operations in parallel_for

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58909 TensorIterator: Reduce serial_for_each static overhead
* #58656 Replace `dim_apply` with `TensorIterator`
* **#58655 TensorIteratorReduce: Avoid tensor operations in parallel_for**

Ref gh-56794

The two pass reduction calls `copy_` and `select` inside a parallel region. The
`copy_` can just be moved outside of the parallel region, but avoiding the
`select` call is more complicated because it's needed to construct the
`TensorIterator`. Instead, I factor out a `serial_for_each` free-function that
just takes pointers and strides. Then manually advance the pointer to the
thread-specific slice of data.

Differential Revision: [D28735330](https://our.internmc.facebook.com/intern/diff/D28735330)

peterbell10 committed:
TensorIteratorReduce: Avoid tensor operations in parallel_for

Ref gh-56794

The two pass reduction calls `copy_` and `select` inside a parallel region. The
`copy_` can just be moved outside of the parallel region, but avoiding the
`select` call is more complicated because it's needed to construct the
`TensorIterator`. Instead, I factor out a `serial_for_each` free-function that
just takes pointers and strides. Then manually advance the pointer to the
thread-specific slice of data.

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIteratorReduce: Avoid tensor operations in parallel_for"


Ref gh-56794

The two pass reduction calls `copy_` and `select` inside a parallel region. The
`copy_` can just be moved outside of the parallel region, but avoiding the
`select` call is more complicated because it's needed to construct the
`TensorIterator`. Instead, I factor out a `serial_for_each` free-function that
just takes pointers and strides. Then manually advance the pointer to the
thread-specific slice of data.

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIteratorReduce: Avoid tensor operations in parallel_for"


Ref gh-56794

The two pass reduction calls `copy_` and `select` inside a parallel region. The
`copy_` can just be moved outside of the parallel region, but avoiding the
`select` call is more complicated because it's needed to construct the
`TensorIterator`. Instead, I factor out a `serial_for_each` free-function that
just takes pointers and strides. Then manually advance the pointer to the
thread-specific slice of data.

[ghstack-poisoned]

ngimel commented:
@peterbell10 can you please run perf counters benchmarks on this? This is a very hot codepath, so any negative changes will be noticeable. For an example benchmark see e.g. here (where btw I'm changing the same code paths) https://github.com/pytorch/pytorch/pull/58810

ngimel commented on aten/src/ATen/native/TensorIteratorReduce.cpp:
```diff
@@ -35,36 +36,32 @@ static bool use_two_pass_reduction(TensorIteratorBase& iter) {
-    }
-  }
+    // Bump output ptr so each thread has its own ouput slice
+    auto base_ptrs = first_reduce.get_base_ptrs();
```
what you are doing here can probably be achieved by `unsafe_replace_operand` method in TensorIterator?

peterbell10 replied:
TBH I didn't know about `unsafe_replace_operand`. However, since it modifies the iterator, I would need to take an extra copy of the `TensorIterator` on each thread. In terms of callgrind, with this PR I see for a 1-element sum() and inclusive counts:
```
  433753  ???:at::two_pass_reduction(at::T ... har**, long const*, long, long)>)
```
and using `unsafe_replace_operand` it's
```
  452233  ???:at::two_pass_reduction(at::T ... har**, long const*, long, long)>)
```

So on the order if 5% overhead. That said, this is only used for large reductions anyway so perhaps that's okay.

ngimel replied:
You are changing `TensorIteratorBase::serial_for_each(loop2d_t loop, Range range)` which is used everywhere, so can you please benchmark a regular unary and binary pointwise operation?

peterbell10 committed:
Update on "TensorIteratorReduce: Avoid tensor operations in parallel_for"


Ref gh-56794

The two pass reduction calls `copy_` and `select` inside a parallel region. The
`copy_` can just be moved outside of the parallel region, but avoiding the
`select` call is more complicated because it's needed to construct the
`TensorIterator`. Instead, I factor out a `serial_for_each` free-function that
just takes pointers and strides. Then manually advance the pointer to the
thread-specific slice of data.

[ghstack-poisoned]

peterbell10 commented:
@ngimel I did see some regressions in the `serial_for_each` instruction count, so I've made some tweaks to where it now out-performs the old version in your `add_` benchmark. Using inclusive counts, `serial_for_each` previously has 18660 and now has 17490.

ngimel commented:
> @ngimel I did see some regressions in the `serial_for_each` instruction count, so I've made some tweaks to where it now out-performs the old version in your `add_` benchmark. Using inclusive counts, `serial_for_each` previously has 18660 and now has 17490.

Wait, but those tweaks in a separate PR (#58810, use base_ptrs directly if range.start is 0, and don't pad strides for 1d) lead to approx 5K improvement, so (18660-17490) is less than could be expected

peterbell10 commented:
> use base_ptrs directly if range.start is 0

Because I'm using `ArrayRef` for base_ptrs, I still need to copy. 

> don't pad strides for 1d

I'm still padding the strides because I'm not sure I agree that it's valid to pass an under-sized array to the loop.  While it's true that value isn't necessary, it's still likely that the loop function will read the value. So technically it's undefined behaviour.

If I apply those two optimizations, I get it down to 15000. I have some other ideas that could bring it down much more though, and if it's okay I'd like to follow up in a separate PR. I just think that would go outsize of the scope of this PR. As it stands, there is an incidental improvement, but it's not the focus here.

ngimel commented:
>I'm still padding the strides because I'm not sure I agree that it's valid to pass an under-sized array to the loop. While it's true that value isn't necessary, it's still likely that the loop function will read the value. So technically it's undefined behaviour.

Existing code does not read the value (there's valgrind test for this, that fired for reductions and 0-dim pw operations)

peterbell10 commented:
Currently, most of the code uses 1d loop functions where this isn't an issue because the 2d-to-1d wrapper happens to not read it.  However, reductions are a key example that use the 2d loop interface and that's why reductions failed and the `is_reduction_` check had to be added in your PR. In my `nonzero` port PR (#58811) I also use a 2d-loop and that would start to invoke undefined behaviour now.

peterbell10 committed:
Update on "TensorIteratorReduce: Avoid tensor operations in parallel_for"


Ref gh-56794

The two pass reduction calls `copy_` and `select` inside a parallel region. The
`copy_` can just be moved outside of the parallel region, but avoiding the
`select` call is more complicated because it's needed to construct the
`TensorIterator`. Instead, I factor out a `serial_for_each` free-function that
just takes pointers and strides. Then manually advance the pointer to the
thread-specific slice of data.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/TensorIteratorReduce.cpp:
```diff
@@ -35,36 +36,32 @@ static bool use_two_pass_reduction(TensorIteratorBase& iter) {
-  std::fill(written.get(), written.get() + max_threads, false);
+  auto buffer_stride = buffer.strides()[0] * buffer.element_size();
+  auto buffer_0 = buffer[0];
+  auto first_reduce = TensorIterator::reduce_op(buffer_0, iter.input(0));
```
The correct operation depends on output not being replaced or resized in TensorIterator, which is true now but can change. Can you assert (or debug assert) that buffer_0 data_ptr and first_reduce.output().data_ptr are the same?

ngimel commented on c10/util/SmallBuffer.h:
```diff
@@ -0,0 +1,53 @@
+namespace c10 {
+
+template <typename T, size_t N>
+class SmallBuffer {
```
This is great, and we should use it more in places we are currently using DimVectors.

ngimel commented:
Small comment, otherwise good to go

ngimel commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -665,29 +666,7 @@ void TensorIteratorBase::serial_for_each(loop2d_t loop, Range range) const {
-  auto base_ptrs = get_base_ptrs();
-  if (ndim() <= 1) {
-    if (range.begin > 0) {
-      auto ptrs = get_data_ptrs(base_ptrs, {range.begin});
```
looks like you no longer need `get_data_ptrs` in TensorIterator.cpp since you have it in TensorIteratorInternal.h?

peterbell10 committed:
Update on "TensorIteratorReduce: Avoid tensor operations in parallel_for"


Ref gh-56794

The two pass reduction calls `copy_` and `select` inside a parallel region. The
`copy_` can just be moved outside of the parallel region, but avoiding the
`select` call is more complicated because it's needed to construct the
`TensorIterator`. Instead, I factor out a `serial_for_each` free-function that
just takes pointers and strides. Then manually advance the pointer to the
thread-specific slice of data.

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIteratorReduce: Avoid tensor operations in parallel_for"


Ref gh-56794

The two pass reduction calls `copy_` and `select` inside a parallel region. The
`copy_` can just be moved outside of the parallel region, but avoiding the
`select` call is more complicated because it's needed to construct the
`TensorIterator`. Instead, I factor out a `serial_for_each` free-function that
just takes pointers and strides. Then manually advance the pointer to the
thread-specific slice of data.

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28735330).
